Improve and unify debug views of dictionaries.

arturek opened the pull request:
The change allows generic and non-generic dictionaries to be displayed in the same way: with separate columns for keys and values with an ability to expand each column.

Currently non-generic dictionaries are displayed using two columns but without an option to expand keys or values: 
![Hashtable debug view](https://github.com/dotnet/runtime/assets/381257/7282ae38-e494-4114-b0b8-966e56d2cb95)

Generic dictionaries are displayed as ordered lists of key/value pairs with a drill-down functionality: 
![Dictionary debug view](https://github.com/dotnet/runtime/assets/381257/62600f46-3fe9-4db5-bcff-0c62ae4df677)

The PR unifies the way both kinds of dictionaries are displayed by a debugger.  Dictionaries are shown using two-column format (which IMO better represents a dictionary) and both keys and values can be expanded:
![New debug view of dictionaries](https://github.com/dotnet/runtime/assets/381257/1fb57417-9604-4c63-ab44-dc0b38d2ef68)

Fixes #88736

arturek committed:
Improves and unifies debug views of dictionaries.

The change alows generic and non-generic dictionaries to be displayed in the same way: with separate columns for keys and values with an ability to expand each column.

Fixes #88736

eiriktsarpalis commented on src/libraries/System.Private.CoreLib/src/System/Collections/Generic/IDictionaryDebugView.cs:
```diff
@@ -17,12 +17,17 @@ public IDictionaryDebugView(IDictionary<K, V> dictionary)
             {
-                KeyValuePair<K, V>[] items = new KeyValuePair<K, V>[_dict.Count];
-                _dict.CopyTo(items, 0);
+                var keyValuePairs = new KeyValuePair<K, V>[_dict.Count];
```
Is the intermediate KeyValuePair array necessary here? Seems it could be avoided.

arturek replied:
The only way I found to avoid creating this temporary array was to enumerate the dictionary. However, such a change slowed down the code by at least 30%. I think this would affect the debugger UI responsiveness more than the array allocation. Have you thought about some other solution?

eiriktsarpalis replied:
Right, I incorrectly assumed this was only targeting `Dictionary<K,V>` and missed it's targeting the interface. Looks good then.

eiriktsarpalis reviewed (commented):
Awesome, thanks! We should investigate and check if other included dictionary types need the same treatment (including the ones in `System.Collections.Immutable` and `System.Collections.Frozen`).

eiriktsarpalis commented:
@arturek this seems to be causing test failures in CI, could you take a look?

arturek commented:
@eiriktsarpalis
Immutable and frozen dictionaries use their own debugger proxy implementation to cache their enumerated items. That proxy displays items the same way as a generic dictionary does it now. I can make updates to them after this PR is completed. Other types of generic dictionaries (in `System.Collections`, `System.Collections.Concurrent`, `System.Collections.NonGeneric` and `System.Private.CoreLib`) are updated by this PR. I hope there are no dictionaries defined in other places.

I'll fix the tests. I thought they are unrelated but now I can see the reason they are failing.

arturek committed:
Merge branch 'main' of https://github.com/dotnet/runtime into DictionaryDebugView

arturek committed:
Merge branch 'main' of https://github.com/dotnet/runtime into DictionaryDebugView

arturek committed:
Fix the DebuggerView tests of dictionaries

Included non-generic dictionaries in tests

arturek committed:
Merge branch 'main' of https://github.com/dotnet/runtime into DictionaryDebugView

arturek committed:
Fix more DebuggerView tests

* Included more types that implement IDictionary in the DebuggerView tests.
* Improved the testing code to support classes with attributes declared by their base classes.
* Fixed .Net Framework 4.8 build error by removing a dependency on the record c# feature.
* Fixed tests remaining tests (outside of the System.Collections subset)

arturek committed:
Merge branch 'main' of https://github.com/dotnet/runtime into DictionaryDebugView

arturek committed:
Fix DebugView.Tests build errors on .Net Framework

eiriktsarpalis commented:
Still seeing a number of failing tests, e.g.
```
  Starting:    System.Collections.Immutable.Tests (parallel test collections = on, max threads = 2)
    System.Collections.Tests.DebugView_Tests.TestDebuggerAttributes_Null(obj: []) [FAIL]
      System.InvalidOperationException : Expected one DebuggerTypeProxyAttribute on System.Collections.ListDictionaryInternal.
      Stack Trace:
        /_/src/libraries/Common/tests/System/Diagnostics/DebuggerAttributes.cs(129,0): at System.Diagnostics.DebuggerAttributes.FindAttribute(Type type, Type attributeType)
        /_/src/libraries/Common/tests/System/Diagnostics/DebuggerAttributes.cs(141,0): at System.Diagnostics.DebuggerAttributes.GetProxyType(Type type, Type[] genericTypeArguments)
        /_/src/libraries/Common/tests/System/Collections/DebugView.Tests.cs(166,0): at System.Collections.Tests.DebugView_Tests.TestDebuggerAttributes_Null(Object obj)
    System.Collections.Tests.DebugView_Tests.TestDebuggerAttributes_Null(obj: [System.Collections.DictionaryEntry, System.Collections.DictionaryEntry]) [FAIL]
      System.InvalidOperationException : Expected one DebuggerTypeProxyAttribute on System.Collections.ListDictionaryInternal.
      Stack Trace:
        /_/src/libraries/Common/tests/System/Diagnostics/DebuggerAttributes.cs(129,0): at System.Diagnostics.DebuggerAttributes.FindAttribute(Type type, Type attributeType)
        /_/src/libraries/Common/tests/System/Diagnostics/DebuggerAttributes.cs(141,0): at System.Diagnostics.DebuggerAttributes.GetProxyType(Type type, Type[] genericTypeArguments)
        /_/src/libraries/Common/tests/System/Collections/DebugView.Tests.cs(166,0): at System.Collections.Tests.DebugView_Tests.TestDebuggerAttributes_Null(Object obj)
    System.Collections.Tests.DebugView_Tests.TestDebuggerAttributes_Dictionary(obj: [], expected: []) [FAIL]
      System.InvalidOperationException : Expected one DebuggerDisplayAttribute on System.Collections.ListDictionaryInternal.
      Stack Trace:
        /_/src/libraries/Common/tests/System/Diagnostics/DebuggerAttributes.cs(129,0): at System.Diagnostics.DebuggerAttributes.FindAttribute(Type type, Type attributeType)
        /_/src/libraries/Common/tests/System/Diagnostics/DebuggerAttributes.cs(115,0): at System.Diagnostics.DebuggerAttributes.ValidateDebuggerDisplayReferences(Object obj)
        /_/src/libraries/Common/tests/System/Collections/DebugView.Tests.cs(140,0): at System.Collections.Tests.DebugView_Tests.TestDebuggerAttributes_Dictionary(IDictionary obj, KeyValuePair`2[] expected)
```

arturek commented:
@eiriktsarpalis:
I'm working on it. The latest issues are quite unexpected.

The tests you mentioned failed because `ListDictionaryInternal` doesn't have `DebuggerTypeProxyAttribute`, but this attribute was [added a month ago](https://github.com/dotnet/runtime/blob/main/src/libraries/System.Private.CoreLib/src/System/Collections/ListDictionaryInternal.cs). They also fail because they use the previous definition of the `IDictionaryDebugView` instead of the one modified by this PR. What's even stranger, these tests pass on my machine when run in the System.Collection solution, but not in System.Collection.Immutable (and earlier I used only the former solution to verify my fixes before committing them).

I'll look at it tomorrow.

arturek committed:
Update DebugView tests to expect different outcomes on .Net Framework

.Net Framwork does not support recent improvements in the way the debugger displays a dictionary. Depending on the framwork used, the debugger view of generic dictionaries and ListDictionaryInternal are different.

arturek commented:
@eiriktsarpalis 
I updated tests and they almost pass. The only failure reason is "_System.Net.Quic.QuicException : The connection timed out from inactivity_" which doesn't seem related to my changes.
I couldn't find out what exactly preprocessor symbols `NETFRAMEWORK` and `NETCOREAPP` represent, and which of them would be better in this case, but current code seems to work as expected.

eiriktsarpalis commented on src/libraries/Common/tests/System/Collections/DebugView.Tests.cs:
```diff
@@ -11,14 +12,87 @@ namespace System.Collections.Tests
     public class DebugView_Tests
     {
-        public static IEnumerable<object[]> TestDebuggerAttributes_Inputs()
+        private static IEnumerable<object[]> TestDebuggerAttributes_GenericDictionaries()
```
Curious, what necessitated the splitting of the test inputs into separate methods?

arturek replied:
Previously the list contained only generic dictionaries and lists, and all of them were displayed and tested the same way. I added non-generic dictionaries which used a different method of testing.  Depending on the .Net version (Framwork or Core) generic dictionaries now behave similar to elements of either the first or the second group. That's why now there are three groups with two testing methods.

eiriktsarpalis commented on src/libraries/Common/tests/System/Collections/DebugView.Tests.cs:
```diff
@@ -29,40 +103,84 @@ public static IEnumerable<object[]> TestDebuggerAttributes_Inputs()
+            DebuggerAttributes.ValidateDebuggerDisplayReferences(obj);
+            DebuggerAttributeInfo info = DebuggerAttributes.ValidateDebuggerTypeProxyProperties(obj);
+            PropertyInfo itemProperty = info.Properties.Single(pr => pr.GetCustomAttribute<DebuggerBrowsableAttribute>().State == DebuggerBrowsableState.RootHidden);
+            var itemArray = itemProperty.GetValue(info.Instance) as Array;
```
Please replace `var` keywords with explicit type names.

arturek replied:
Done.

eiriktsarpalis commented on src/libraries/Common/tests/System/Diagnostics/DebuggerAttributes.cs:
```diff
@@ -86,18 +93,52 @@ public static IEnumerable<PropertyInfo> GetDebuggerVisibleProperties(Type debugg
+                {
+                    if (attributes.Length > 1)
+                    {
+                        throw new InvalidOperationException($"Expected one {attributeType.Name} on {type} but found more.");
```
Should this be an assert instead?

arturek replied:
IMO helper methods such as this one should not determine the outcome of a test (unless it's their main purpose). Test methods can handle exceptions (or not) based on their requirements. These tests worked this way previously, I haven't change it.

eiriktsarpalis commented on src/libraries/System.Collections.NonGeneric/tests/SortedListTests.cs:
```diff
@@ -223,19 +223,21 @@ public void DebuggerAttribute_NormalList()
-            object[] items = (object[])infoProperty.GetValue(debuggerAttribute.Instance);
-            Assert.Equal(list.Count, items.Length);
+            PropertyInfo itemProperty = debuggerAttribute.Properties.Single(pr => pr.GetCustomAttribute<DebuggerBrowsableAttribute>().State == DebuggerBrowsableState.RootHidden);
+            var itemArray = itemProperty.GetValue(debuggerAttribute.Instance) as Array;
```
var -> type name please.

arturek replied:
Done.

eiriktsarpalis commented on src/libraries/Common/tests/System/Diagnostics/DebuggerAttributes.cs:
```diff
@@ -86,18 +93,52 @@ public static IEnumerable<PropertyInfo> GetDebuggerVisibleProperties(Type debugg
-                .Where(a => a.AttributeType == typeof(DebuggerTypeProxyAttribute))
-                .ToArray();
-            if (attrs.Length != 1)
+            for (var t = type; t != null; t = t.BaseType)
```
var -> type name please

arturek replied:
Done.

eiriktsarpalis commented on src/libraries/System.ObjectModel/tests/ReadOnlyDictionary/ReadOnlyDictionaryTests.cs:
```diff
@@ -239,8 +239,8 @@ public static void DebuggerAttributeTests()
             PropertyInfo itemProperty = info.Properties.Single(pr => pr.GetCustomAttribute<DebuggerBrowsableAttribute>().State == DebuggerBrowsableState.RootHidden);
-            KeyValuePair<int, int>[] pairs = itemProperty.GetValue(info.Instance) as KeyValuePair<int, int>[];
-            Assert.Equal(dict, pairs);
+            var itemArray = itemProperty.GetValue(info.Instance) as Array;
```
var -> type name

arturek replied:
Done.

eiriktsarpalis commented on src/libraries/System.Private.CoreLib/src/System/Collections/Generic/DebugViewDictionaryItem.cs:
```diff
@@ -0,0 +1,32 @@
+    /// Defines a key/value pair for displaying an item of a dictionary by a debugger.
+    /// </summary>
+    [DebuggerDisplay("{Value}", Name = "[{Key}]")]
+    internal readonly struct DebugViewDictionaryItem<K, V>
```
Generic naming conventions:
```suggestion
    internal readonly struct DebugViewDictionaryItem<TKey, TValue>
```

arturek replied:
Done.

eiriktsarpalis commented on src/libraries/System.Private.CoreLib/src/System/Collections/Generic/IDictionaryDebugView.cs:
```diff
@@ -17,12 +17,17 @@ public IDictionaryDebugView(IDictionary<K, V> dictionary)
 
         [DebuggerBrowsable(DebuggerBrowsableState.RootHidden)]
-        public KeyValuePair<K, V>[] Items
+        public DebugViewDictionaryItem<K, V>[] Items
```
Any chance you could rename the type parameters to be `TKey` and `TValue` here as well?

arturek replied:
Done.

arturek committed:
Applied suggested changes and fixes

* mostly code style changes
* restored a rest for an empty HashSet.
* fixed testing of the generic SortedList.

arturek commented on src/libraries/Common/tests/System/Collections/DebugView.Tests.cs:
```diff
@@ -11,14 +12,87 @@ namespace System.Collections.Tests
+        private static IEnumerable<object[]> TestDebuggerAttributes_ListInputs()
         {
-            yield return new object[] { new Dictionary<int, string>() };
-            yield return new object[] { new HashSet<string>() };
```
Thanks. I lost this line while splitting the list.  Fixed.
I checked again whether it was the only missing element and I added one more test case for a non-empty generic SortedList.

eiriktsarpalis reviewed (commented):
Almost forgot, but would it be possible to include this change in immutable collections as well? If you prefer to do this in a future PR that's also fine.

stephentoub commented on src/libraries/System.Private.CoreLib/src/System/Collections/Generic/DebugViewDictionaryItem.cs:
```diff
@@ -0,0 +1,32 @@
+            Value = keyValue.Value;
+        }
+
+        [DebuggerBrowsable(DebuggerBrowsableState.Collapsed)]
+        public TKey Key { get; init; }
+
+        [DebuggerBrowsable(DebuggerBrowsableState.Collapsed)]
+        public TValue Value { get; init; }
```
```suggestion
        [DebuggerBrowsable(DebuggerBrowsableState.Collapsed)]
        public TKey Key { get; }

        [DebuggerBrowsable(DebuggerBrowsableState.Collapsed)]
        public TValue Value { get; }
```

arturek replied:
Done

stephentoub commented on src/libraries/System.Private.CoreLib/src/System/Collections/Hashtable.cs:
```diff
@@ -1385,7 +1382,7 @@ public override void OnDeserialization(object? sender)
             }
 
-            internal override KeyValuePairs[] ToKeyValuePairsArray()
+            internal override DebugViewDictionaryItem<object, object?>[] ToKeyValuePairsArray()
```
Nit: this is no longer a "KeyValuePairsArray"

arturek replied:
Done.

arturek committed:
Minor improvents

Renamed an internal method to match its new behavior and removed unnecessary init accessors.

arturek commented:
@eiriktsarpalis 
> Almost forgot, but would it be possible to include this change in immutable collections as well? If you prefer to do this in a future PR that's also fine.

It's on my list. I'd prefer to do it in a separate PR just to close this phase and keep scope of this PR smaller.

eiriktsarpalis reviewed (approved):
Thanks!
